src/basic/in-addr-util.c

```diff
@@ -226,6 +226,70 @@ int in_addr_prefix_next(int family, union in_addr_union *u, unsigned prefixlen)
         return -EAFNOSUPPORT;
 }
 
+int in_addr_prefix_nth(int family, union in_addr_union *u, unsigned prefixlen, unsigned nth) {
+        assert(u);
+
+        if (prefixlen <= 0)
+                return 0;
+
+        if (nth == 0) { // identity??
+                return 1;
+        }
+
+        if (family == AF_INET) {
+                uint32_t c, n;
+                if (prefixlen > 32)
+                        prefixlen = 32;
+
+                c = be32toh(u->in.s_addr);
+
+                n = c + (nth * (1UL << (32 - prefixlen)));
+
+                if (n < c)
+                        return 0; // overflow
+
+                n &= 0xFFFFFFFFUL << (32 - prefixlen);
+                u->in.s_addr = htobe32(n);
+                return 1;
+
+        }
+
+        if (family == AF_INET6) {
+                struct in6_addr result = {};
+                uint8_t overflow = 0;
+                uint64_t delta ; // this assumes that we only ever have to up to 2<<64 subnets
+                unsigned i;
+                unsigned start_byte = (prefixlen -1) / 8;
+
+                if (prefixlen > 128)
+                        prefixlen = 128;
+
+                /* First calculate what we have to add */
+                delta = nth << ((128 - prefixlen) % 8);
+
+                for (i = 16; i > 0; i--) {
+                        unsigned j = i - 1;
+                        unsigned d = 0;
+
+                        if (j <= start_byte) {
+                                d = delta & 0xFF;
+                                delta >>= 8;
+                                result.s6_addr[j] = u->in6.s6_addr[j] + d + overflow;
+                                overflow = (result.s6_addr[j] < u->in6.s6_addr[j]);
```

Comment:
as above. 

(we still have some oveflow checks like this in our codebase, but for new code we prefer doing it correctly, i.e. beforehand)